src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -1097,12 +1093,32 @@ void CodeGen::genAVXIntrinsic(GenTreeHWIntrinsic* node)
 {
     NamedIntrinsic intrinsicID = node->gtHWIntrinsicId;
     var_types      baseType    = node->gtSIMDBaseType;
-    instruction    ins         = INS_invalid;
+    emitAttr       attr        = EA_ATTR(node->gtSIMDSize);
+    var_types      targetType  = node->TypeGet();
+    instruction    ins         = Compiler::insOfHWIntrinsic(intrinsicID, baseType);
+    GenTree*       op1         = node->gtGetOp1();
+    GenTree*       op2         = node->gtGetOp2();
+    regNumber      targetReg   = node->gtRegNum;
+    emitter*       emit        = getEmitter();
 
     genConsumeOperands(node);
 
     switch (intrinsicID)
     {
+        case NI_AVX_SetZeroVector256:
+        {
+            assert(op1 == nullptr);
+            assert(op2 == nullptr);
+            if (!compiler->compSupports(InstructionSet_AVX2) && varTypeIsIntegral(baseType))
```

Comment:
I don't quite understand what's this `if` for and what the whole thing has to do with this PR. Can you clarify?

Reply:
`SetZeroVector256` will generate `pxor` with integral base-typ, but `pxor` is a AVX2 instruction, so we generate `xorps` on AVX machines to avoid the data bypass penalty.

Reply:
@fiigii, given that AVX doesn't expose any integer operations anyways, might it be better to only expose the integer based `SetZeroVector256` on the AVX2 isa (I believe this is what we did for SSE/SSE2)?

Reply:
This would also ensure users are explicitly aware of the penalty they may receive.

Reply:
SSE/SSE2 is somewhat simpler, and weirder at the same time. SSE ended up with a non generic `SetZeroVector128` that gives you a `Vector128<float>`. SSE2 has a generic `SetZeroVector128<T>()` that gives you exactly what? I would guess it won't give you a `Vector128<float>` because that's the job of SSE.

If you really want to have an exacting mapping between intrinsic methods and CPU instructions I'd avoid using generic methods. It ends up being just weird and not very helpful to the developer who's supposed to use these intrinsic.

Reply:
Though as a developer I'd expect `new Vector128<float>()` or `default` to just work and emit the right code. Having to use something like `Sse2.SetZeroVector128<int>()` just to get a 0 is pretty obnoxius.

Reply:
> If you really want to have an exacting mapping between intrinsic methods and CPU instructions I'd avoid using generic methods.

We've actually done a pretty good job on that so far (at least I think). For the most part, we only use the generic methods when all 10 supported primitives (float, double, byte, sbyte, short, ushort, int, uint, long, and ulong) are supported for that intrinsic in that ISA.

We have been "exploding" the APIs in the cases where only a subset of the supported primtives are supported by that ISA.

I would guess that `SetZeroVector128` was just missed (this also doesn't use the naming convention we have been following, and should be called `SetZeroVector128Single`, etc).

CC. @eerhardt, @fiigii, @CarolEidt

Reply:
> Though as a developer I'd expect new Vector128<float>() or default to just work and emit the right code.

This should work (you should get zero), but it wouldn't use `xorps`, `xorpd`, or `pxor` unless the JIT was already configured to zero structs that way.

Reply:
> but it wouldn't use xorps, xorpd, or pxor 

Well, it should. I suspect at the moment it will always generate `xorps` which is good enough.

Reply:
> We have been "exploding" the APIs in the cases where only a subset of the supported primtives are supported by that ISA.

Reply:
> I would guess that SetZeroVector128 was just missed (this also doesn't use the naming convention we have been following, and should be called SetZeroVector128Single, etc).

Ah, I think we should move `Sse2.SetZeroVector128<T>` to `Sse` to match the AVX API.

> given that AVX doesn't expose any integer operations anyways, might it be better to only expose the integer based SetZeroVector256 on the AVX2 isa (I believe this is what we did for SSE/SSE2)?

I do not think it is necessary, for we never promise the generated instructions of helper-intrinsics and the codgen may be changed due to ISA development. One example is `SetAllVector128/256<T>` can generate `broadcast` on AVX2 machines.

Reply:
Logged at https://github.com/dotnet/coreclr/issues/16575.